Update sanitizer tool to clean messages more completely

heliocliu opened the pull request:
#633
(previous PR #3354 targeted the old repo and got closed so just opening a new one)

The sanitize tool is a bit incomplete with what it scrubs out.

-Add JSON schema definitions for all the message types we expect to encounter.  Messages are validated against these schema to ensure we don't hit unexpected messages we don't know how to handle (main use here is to check that there aren't extra unexpected fields)
-Add stable replacement of known PII fields (e.g. the same email address gets the same replacement every time)
-By default attempt to scrub in a way that still makes the document readable (PENDING FOR TEST), and add options to do a full scrub and to continue on unexpected messages instead of failing.
1. Full scrub is not expected to be readable and removes things like data store names
2. NOT FULLY SUPPORTED: Some data-store specific props may still need to be special cased so they work correctly.  I don't think it's possible to guarantee properties essential to the component is not removed in these cases because of how component-specific values are just an object that doesn't match a specific schema

-I used the fluid readout document to test.  The replaced document reads into the replay-tool (with the same output, including errors).  It does not work with the debugger tool in fluid preview (this scenario appears broken for snapshots right now and doesn't appear to have ever been supported for message.json files).  The debugger tool scenario will be tracked by separate issues.

heliocliu committed:
initial changes

includes changes from previous branch plus some others

heliocliu committed:
chunked op support changes

-fix chunked op pending processing check
-do replacement for root-level chunkedops
-do writeback for replaced chunked op contents

heliocliu committed:
add support for nested chunkedOp

heliocliu committed:
fix policy check

heliocliu commented on packages/test/snapshots/src/messageSchema.ts:
```diff
@@ -0,0 +1,433 @@
+    type: "null",
+};
+
+export const joinDataSchema = {
```
Some of these schema were difficult to get from the code and involved examination of a real world document... Might be worthwhile for anyone that writes messages to make this more clear where it's not already?

anthony-murphy commented on packages/test/snapshots/src/sanitize.ts:
```diff
@@ -3,124 +3,692 @@
+        }
+
+        if (type === TextType.Generic) {
+            return "xxxxx";
```
> return "xxxxx"; [](start = 11, length = 16)

this will break the merge tree. you need to replace with the same number of existing characters.

anthony-murphy replied:
in general i think you should always trying to replace strings with like strings. i would break things by word using regex, and map all words to the same value across everything you sanitize to attempt to preserve any hidden invariants as well.

---
In reply to: [500441391](https://github.com/microsoft/FluidFramework/pull/3786#discussion_r500441391) [](ancestors = 500441391)

heliocliu replied:
Changed all replacement to preserve length and for words to be replaced 1:1.  Clarified in the header that this reduces how "sanitized" the document actually is.  replay-tool reads the documents fine with these changes, but debugger tool doesn't.  debugger tool appears to have other issues/feature gaps that prevents this though, so I'll open other issues to track that.  I don't think it would be reasonable to cover debugger tool/fluid preview support in this PR.

anthony-murphy commented on packages/test/snapshots/src/sanitize.ts:
```diff
@@ -3,124 +3,692 @@
+                break;
+            default:
+                return "xxxxx";
         }
```
same here

anthony-murphy commented on packages/test/snapshots/src/sanitize.ts:
```diff
@@ -3,124 +3,692 @@
+                replacement = `email_${uuid()}}`;
+                break;
+            case TextType.Name:
+                replacement = `name_${uuid()}`;
```
i'm not sure there is much value in preserving email and name

anthony-murphy replied:
you should also match char length her

---
In reply to: [500444976](https://github.com/microsoft/FluidFramework/pull/3786#discussion_r500444976) [](ancestors = 500444976)

heliocliu committed:
iteration updates

-preserve word token lengths and use random strings instead of 'xxxxx'
-update replaced chunkedop size to match the original message because chunked ops will no longer get smaller

heliocliu committed:
Merge branch 'main' into SanitizerUpdates

heliocliu committed:
Update sanitize.ts

fix new lint errors
fix replacement of snapshot kvs

anthony-murphy reviewed (approved):
:shipit:
